### PR TITLE
Add documentation regarding neural type system.

### PR DESCRIPTION
**Pull request by okuchaiev**
Introduces "add_port_docs" which automatically generates docstrings for input and output ports.
Replaces old neural type tutorial with a new one.

### CODE REVIEW

File: nemo/backends/pytorch/common/losses.py [hunk 1]
```diff
@@ -32,18 +33,16 @@ class SequenceLoss(LossNM):
     """
 
     @property
+    @add_port_docs()
```
Comment (tkornuta-nvidia) [inline nemo/backends/pytorch/common/losses.py]: a decorator! nice! 👍 :)

File: nemo/backends/pytorch/nm.py [hunk 2]
```diff
@@ -268,6 +268,7 @@ def num_weights(self):
 
     @abstractmethod
     def __len__(self):
+        """Dataset size"""
```
Comment (blisc) [inline nemo/backends/pytorch/nm.py]: You doubled up on this comment

File: nemo/core/neural_types/neural_type.py [hunk 3]
```diff
@@ -73,6 +78,19 @@ def __init__(self, axes: Optional[Tuple] = None, elements_type: ElementType = Vo
             self.axes = None
         self.optional = optional
 
+    def __repr__(self) -> str:
```
Comment (blisc) [inline nemo/core/neural_types/neural_type.py]: NeuralType already has a str(). Should we be printing this or str()?

File: nemo/backends/pytorch/nm.py (hunk 2 above)
Comment (okuchaiev) [reply nemo/backends/pytorch/nm.py]: fixed

File: nemo/core/neural_types/neural_type.py (hunk 3 above)
Comment (okuchaiev) [reply nemo/core/neural_types/neural_type.py]: good catch - removed that, since for now str() and repr() should be the same

**Comment by okuchaiev** [comment]
> Can you either update changelog or the developer guide? Preferably both

updated both

**Comment by blisc** [comment]
Please fix CI but otherwise looks good to me

**Pull request merged**